

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>READemption’s subcommands &mdash; READemption 0.5.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Performing an example analysis" href="example_analysis.html" />
    <link rel="prev" title="Docker image" href="docker_image.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> READemption
          

          
            
            <img src="_static/READemption_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation and updating</a></li>
<li class="toctree-l1"><a class="reference internal" href="live_and_installation_image.html">Live system and installation image</a></li>
<li class="toctree-l1"><a class="reference internal" href="docker_image.html">Docker image</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">READemption’s subcommands</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#create">create</a></li>
<li class="toctree-l2"><a class="reference internal" href="#align">align</a></li>
<li class="toctree-l2"><a class="reference internal" href="#coverage">coverage</a></li>
<li class="toctree-l2"><a class="reference internal" href="#gene-quanti">gene_quanti</a></li>
<li class="toctree-l2"><a class="reference internal" href="#deseq">deseq</a></li>
<li class="toctree-l2"><a class="reference internal" href="#viz-align">viz_align</a></li>
<li class="toctree-l2"><a class="reference internal" href="#viz-gene-quanti">viz_gene_quanti</a></li>
<li class="toctree-l2"><a class="reference internal" href="#viz-deseq">viz_deseq</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="example_analysis.html">Performing an example analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="versions.html">Versions/Change log</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">READemption</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>READemption’s subcommands</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="reademption-s-subcommands">
<h1>READemption’s subcommands<a class="headerlink" href="#reademption-s-subcommands" title="Permalink to this headline">¶</a></h1>
<p>In general the subcommands need at least one argument - the analysis
folder. If this is not given READemption assumes that the current
folder is the analysis folder.</p>
<div class="section" id="create">
<h2>create<a class="headerlink" href="#create" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">create</span></code> generates the required folder structure for input and
output files. Once these folders are created the input files have to
be placed into the correct locations. As a minimal requirement,
RNA-Seqs reads in FASTA format (can be compressed with <code class="docutils literal notranslate"><span class="pre">bzip2</span></code> or
<code class="docutils literal notranslate"><span class="pre">gzip</span></code>) must be placed in <code class="docutils literal notranslate"><span class="pre">input/reads</span></code> and the reference sequence
in FASTA format must be copied or linked in
<code class="docutils literal notranslate"><span class="pre">input/reference_sequences</span></code>. For the command <code class="docutils literal notranslate"><span class="pre">gene_quanti</span></code>
annotation files in GFF3 format have to be put into
<code class="docutils literal notranslate"><span class="pre">input/annotations</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">reademption</span> <span class="n">create</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="n">project_path</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>  <span class="n">Name</span><span class="o">/</span><span class="n">path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>    <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
</pre></div>
</div>
</div>
<div class="section" id="align">
<h2>align<a class="headerlink" href="#align" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">align</span></code> performs the clipping and size filtering of the reads, as
well as the actual aligning to the reference sequences. It also
generates statistics about the steps (e.g. number of aligned reads,
number of mappings). As the result of this steps are needed by the
other subcommands it has to be run before the others. It requires
reads in FASTA or FASTQ format (or counterparts compressed with
<code class="docutils literal notranslate"><span class="pre">gzip</span></code> or <code class="docutils literal notranslate"><span class="pre">bzip2</span></code>) and reference sequences in FASTA
format. <code class="docutils literal notranslate"><span class="pre">align</span></code> generates the read alignments in BAM format
(<code class="docutils literal notranslate"><span class="pre">*.bam</span></code>) and also index files for those (<code class="docutils literal notranslate"><span class="pre">*.bam.bai</span></code>). Is also
stores unmapped reads so that they can be inspected e.g. to search for
contaminations. The file
<code class="docutils literal notranslate"><span class="pre">output/align/reports_and_stats/read_alignment_stats.csv</span></code> lists
several mapping statistics. The folder
<code class="docutils literal notranslate"><span class="pre">output/align/reports_and_stats/stats_data_json/</span></code> contains files
with the original countings in JSON format. Please be aware that
READemption can perform only basic quality trimming and adapter
clipping. If this is not sufficient you can use the <a class="reference external" href="http://hannonlab.cshl.edu/fastx_toolkit/">FASTX toolkit</a>, <a class="reference external" href="https://code.google.com/p/cutadapt/">cutadapt</a> or other tools for the
preprocessing.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>usage: reademption align [-h] [--min_read_length MIN_READ_LENGTH]
                         [--processes PROCESSES]
                         [--segemehl_accuracy SEGEMEHL_ACCURACY]
                         [--segemehl_evalue SEGEMEHL_EVALUE]
                         [--segemehl_bin SEGEMEHL_BIN] [--paired_end]
                         [--split] [--poly_a_clipping] [--realign]
                         [--keep_original_alignments] [--lack_bin LACK_BIN]
                         [--fastq] [--check_for_existing_files] [--progress]
                         [--crossalign_cleaning CROSSALIGN_CLEANING_STRING]
                         [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --min_read_length MIN_READ_LENGTH, -l MIN_READ_LENGTH
                        Minimal read length after clipping (default 12).
                        Should be higher for eukaryotic species.
  --processes PROCESSES, -p PROCESSES
                        Number of processes that should be used (default 1).
  --segemehl_accuracy SEGEMEHL_ACCURACY, -a SEGEMEHL_ACCURACY
                        Segemehl&#39;s minimal accuracy (in %) (default 95).
  --segemehl_evalue SEGEMEHL_EVALUE, -e SEGEMEHL_EVALUE
                        Segemehl&#39;s maximal e-value (default 5.0).
  --segemehl_bin SEGEMEHL_BIN, -s SEGEMEHL_BIN
                        Segemehl&#39;s binary path (default &#39;segemehl.x&#39;).
  --paired_end, -P      Use this if reads are originating from a paired-end
                        sequencing. The members of a pair must be marked with
                        &#39;_p1&#39; and &#39;_p2&#39; in front of the file type suffixes
                        (e.g. &#39;my_sample_p1.fa&#39; and &#39;my_sample_p2.fa&#39; or
                        &#39;my_sample_p1.fa.bz2&#39; and &#39;my_sample_p2.fa.bz2&#39;). This
                        option cannot be use with polyA tail clipping.
  --split, -S           Run segemehl with read splitting.
  --poly_a_clipping, -c
                        Perform polyA tail clipping. This option cannot be
                        used for paired-end reads.
  --realign, -r         Perform realignment of unmapped reads using &#39;lack&#39;.
  --keep_original_alignments, -k
                        Only used with --realign/-r. Keep the alignment file
                        of the primary mapper (segemehl) and the realigner
                        (lack) after merging.
  --lack_bin LACK_BIN, -L LACK_BIN
                        Lack&#39;s binary path (default &#39;lack.x&#39;).
  --fastq, -q           Input reads are in FASTQ not FASTA format.
  --min_phred_score MIN_PHRED_SCORE, -Q MIN_PHRED_SCORE
                        Minimal Phred score. Works only if read are given in
                        FASTQ format. As soon as a based drop below this value
                        it and all the nucleotides downstream of it will be
                        trimmed off.
  --adapter ADAPTER, -A ADAPTER
                        Adapter sequence. If it is found in a read it and all
                        the nucleotides downstream will be trimmed off.
  --check_for_existing_files, -f
                        Check for existing files (e.g. from a interrupted
                        previous run) and do not overwrite them if they exits.
                        Attention! You have to take care that there are no
                        partially generated files left!
  --reverse_complement, -R
                        Map reverse complement of the input reads.
  --progress, -g          Show progress of the segemehl mapping.
  --crossalign_cleaning CROSSALIGN_CLEANING_STRING, -x CROSSALIGN_CLEANING_STRING
                        Remove reads that are cross-mapped to replicons of
                        different species. To associated species and replicons
                        give a string in the following format: &#39;&lt;ORG_NAME_1&gt;:&lt;
                        org_1_repl1&gt;,&lt;org_1_repl2&gt;,..,&lt;org_1_repl_n&gt;;&lt;ORG_NAME
                        _2&gt;:&lt;org_2_repl1&gt;,&lt;org_2_repl2&gt;,..,&lt;org_2_repl_n&gt;&#39;
</pre></div>
</div>
</div>
<div class="section" id="coverage">
<h2>coverage<a class="headerlink" href="#coverage" title="Permalink to this headline">¶</a></h2>
<p><cite>coverage</cite> generates strand specific coverage files in <a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">wiggle format</a> based on the
read alignments. These wiggle files can be viewed in common genome
browser like the <a class="reference external" href="http://bioviz.org/">Integrated genome browser (IGB)</a> or the <a class="reference external" href="https://www.broadinstitute.org/software/igv/">Integrative genome viewer (IGV)</a>. Three sets of wiggle
files will be generated: raw counting values without normalization
(located in the folder <cite>coverage-raw</cite>), normalized by the total number
of aligned reads (abbreviated as tnoar) and the multiplied by the
lowest number of aligned reads of all considered libraries (in folder
<cite>coverage-tnoar_min_normalized</cite>) as well as normalized by the total
number of aligned reads and multiplied by one million
(<cite>coverage-tnoar_mil_normalized</cite>). The different normalizations make a
visual semi-quantitative comparative possible and enable to perform
transcription start site analysis (e.g. using tools like <a class="reference external" href="http://www-ps.informatik.uni-tuebingen.de/itNew/?page_id=1860">TSSPredator</a>). For
each library and set there will be coverage files for the forward and
the reverse strand. The coverages for the forward strand have positive
values while the one for the reverse stand have negative values in
order to make a visual discrimination easy. Per default all reads and
each position of them will be considered. To calculate the coverages
only based on uniquely aligned read use the <code class="docutils literal notranslate"><span class="pre">--unique_only</span></code>
parameter. If only the first base should be considered add
<code class="docutils literal notranslate"><span class="pre">--first_base_only</span></code>. Reads are aligned to multiple location will
account only in fraction to the values of the different positions. For
example a read that is mapped to three different location will
contribute a value of 1/3 to each of the nucleotiedes of these
positions. To turn off this behavior use
<code class="docutils literal notranslate"><span class="pre">--skip_read_count_splitting</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>usage: reademption coverage [-h] [--unique_only] [--normalize_by_uniquely]
                            [--processes PROCESSES]
                            [--skip_read_count_splitting] [--first_base_only]
                            [--check_for_existing_files]
                            [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --unique_only, -u     Use uniquely aligned reads only.
  --normalize_by_uniquely, -U
                        Normalize by the number of uniquely aligned reads. By
                        default the normalization is done based on the total
                        number of aligned reads even if only uniquely aligned
                        reads are used for the coverage calculation.
  --processes PROCESSES, -p PROCESSES
                        Number of processes that should be used (default 1).
  --skip_read_count_splitting, -s
                        Do not split the read counting between different
                        alignings. Default is to do the splitting.
  --non_strand_specific, -d
                        Do not distict between the coverage of the forward and
                        reverse strand but sum them to a single value for each
                        base.
  --first_base_only, -b
                        Only the first bases 5&#39; base of each read aligning is
                        taken into account.
  --check_for_existing_files, -f
                        Check for existing files (e.g. from a interrupted
                        previous run) and do not overwrite them if they exits.
                        Attention! You have to take care that there are no
                        partially generated files left!
</pre></div>
</div>
</div>
<div class="section" id="gene-quanti">
<h2>gene_quanti<a class="headerlink" href="#gene-quanti" title="Permalink to this headline">¶</a></h2>
<p>With <code class="docutils literal notranslate"><span class="pre">gene_quanti</span></code> the number of reads overlapping with each of the
annotation entries is counted and the results are combined in
tables. At least one GGF3 file with annotations has to be placed in
<code class="docutils literal notranslate"><span class="pre">input/annotations</span></code>. The sequence ID of the sequenced must be
precisely the same as the IDs used in the reference sequence FASTA
files. To specify the feature classes (the third column in the GFF3
file e.g. CDS, gene, rRNA, tRNA) that should be quantified the
parameter <code class="docutils literal notranslate"><span class="pre">--features</span></code> can be used. Otherwise countings for all
annotation entries are generated. Per default sense and anti-sense
overlaps are counted and separately listed.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>usage: reademption gene_quanti [-h] [--min_overlap MIN_OVERLAP]
                               [--no_count_split_by_alignment_no]
                               [--no_count_splitting_by_gene_no]
                               [--skip_antisense] [--processes PROCESSES]
                               [--features ALLOWED_FEATURES] [--unique_only]
                               [--pseudocounts] [--check_for_existing_files]
                               [project_path]

positional arguments:
  project_path          Path of the project folder. If none is given the
                        current directory is used.

optional arguments:
  -h, --help            show this help message and exit
  --min_overlap MIN_OVERLAP, -o MIN_OVERLAP
                        Minimal read-annotation-overlap (in nt) (default 1).
  --no_count_split_by_alignment_no, -n
                        Do not split read countings by the number of
                        alignments a read has. By default this count splitting
                        is performed.
  --no_count_splitting_by_gene_no, -l
                        Do not split read countings by the number of genes it
                        overlaps with. By default this count splitting is
                        performed.
  --skip_antisense, -a  Do not count anti-sense read-gene-overlaps. By default
                        sense and anti-sense overlaps are counted and
                        separately reported.
  --non_strand_specific
                        Use countings of reads overlapping with a gene on both
                        strands and sum them up.
  --processes PROCESSES, -p PROCESSES
                        Number of processes that should be used (default 1).
  --features ALLOWED_FEATURES, -t ALLOWED_FEATURES
                        Comma separated list of features that should be
                        considered (e.g. gene, cds, region, exon). Other
                        feature will be skipped. If not specified all features
                        will be considered.
  --unique_only, -u     Use uniquely aligned reads only.
  --pseudocounts, -c    Add a pseudocount of 1 to each gene.
  --check_for_existing_files, -f
                        Check for existing files (e.g. from a interrupted
                        previous run) and do not overwrite them if they exits.
                        Attention! You have to take care that there are no
                        partially generated files left!
</pre></div>
</div>
</div>
<div class="section" id="deseq">
<h2>deseq<a class="headerlink" href="#deseq" title="Permalink to this headline">¶</a></h2>
<p>Differential gene expression can be performed using <code class="docutils literal notranslate"><span class="pre">deseq</span></code> which
will run a <a class="reference external" href="http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a>
analyses for all possible combinations of conditions. To allocated the
conditions to the libraries use the <code class="docutils literal notranslate"><span class="pre">--libs</span></code> and <code class="docutils literal notranslate"><span class="pre">--conditions</span></code>
parameters (e.g. <code class="docutils literal notranslate"><span class="pre">--libs</span>
<span class="pre">SamA_R1.fa,SamA_R2.fa,SamB_R1.fa,SamB_R2.fa</span> <span class="pre">--conditions</span>
<span class="pre">SamA,SamA,SamB,SamB</span></code>).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">reademption</span> <span class="n">deseq</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">--</span><span class="n">libs</span> <span class="n">LIBS</span> <span class="o">--</span><span class="n">conditions</span> <span class="n">CONDITIONS</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">cooks_cutoff_off</span><span class="p">]</span>
                         <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>          <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span> <span class="n">the</span>
                        <span class="n">current</span> <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">libs</span> <span class="n">LIBS</span><span class="p">,</span> <span class="o">-</span><span class="n">l</span> <span class="n">LIBS</span>  <span class="n">Comma</span> <span class="n">separated</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">libraries</span><span class="o">.</span>
  <span class="o">--</span><span class="n">conditions</span> <span class="n">CONDITIONS</span><span class="p">,</span> <span class="o">-</span><span class="n">c</span> <span class="n">CONDITIONS</span>
                        <span class="n">Comma</span> <span class="n">separated</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">same</span> <span class="n">order</span> <span class="k">as</span>
                        <span class="n">their</span> <span class="n">corresponding</span> <span class="n">libraries</span><span class="o">.</span>
  <span class="o">--</span><span class="n">cooks_cutoff_off</span><span class="p">,</span> <span class="o">-</span><span class="n">k</span>
</pre></div>
</div>
</div>
<div class="section" id="viz-align">
<h2>viz_align<a class="headerlink" href="#viz-align" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">viz_align</span></code> plots histograms of the read length distributions of the
reads before and after the read clipping.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">reademption</span> <span class="n">viz_align</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>  <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span> <span class="n">the</span> <span class="n">current</span>
                <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>    <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
</pre></div>
</div>
</div>
<div class="section" id="viz-gene-quanti">
<h2>viz_gene_quanti<a class="headerlink" href="#viz-gene-quanti" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">viz_gene_quanti</span></code> creates scatterplots in which the raw gene wise
quantification values are compared for each library pair
(all-against-all). For each comparison the <a class="reference external" href="https://en.wikipedia.org/wiki/Pearson_product-moment_correlation_coefficient">pearson correllation</a>
(<cite>r</cite>) coefficiant is. Additionally, bar charts that visualize the
distribution of the read counting of the different annotation classes
are plotted.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">reademption</span> <span class="n">viz_gene_quanti</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>  <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span> <span class="n">the</span> <span class="n">current</span>
                <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>    <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
</pre></div>
</div>
</div>
<div class="section" id="viz-deseq">
<h2>viz_deseq<a class="headerlink" href="#viz-deseq" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">viz_deseq</span></code> generates MA-plots of the comparison (log2 fold changes
vs. the base mean) as well as volcano plots (log2 fold changes
vs. p-values / adjusted p-values).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">reademption</span> <span class="n">viz_deseq</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="n">project_path</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">project_path</span>  <span class="n">Path</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">folder</span><span class="o">.</span> <span class="n">If</span> <span class="n">none</span> <span class="ow">is</span> <span class="n">given</span> <span class="n">the</span> <span class="n">current</span>
                <span class="n">directory</span> <span class="ow">is</span> <span class="n">used</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>    <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
<span class="o">--</span><span class="n">max_pvalue</span> <span class="n">MAX_PVALUE</span>
                        <span class="n">Maximum</span> <span class="n">adjusted</span> <span class="n">p</span><span class="o">-</span><span class="n">value</span> <span class="k">for</span> <span class="n">genes</span> <span class="n">considered</span> <span class="n">to</span> <span class="n">be</span>
                        <span class="n">regulated</span><span class="o">.</span> <span class="n">Genes</span> <span class="k">with</span> <span class="n">adjusted</span> <span class="n">p</span><span class="o">-</span><span class="n">values</span> <span class="n">below</span> <span class="n">will</span> <span class="n">be</span>
                        <span class="n">marked</span> <span class="n">red</span><span class="o">.</span> <span class="p">(</span><span class="n">default</span> <span class="mf">0.05</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="example_analysis.html" class="btn btn-neutral float-right" title="Performing an example analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="docker_image.html" class="btn btn-neutral float-left" title="Docker image" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Konrad U. Förstner and Till Sauerwein

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
    <div class="footer">
    This documention is licensed under the <a href="https://creativecommons.org/licenses/by/3.0/">Creative Commons Attribution license</a>, 2019, Konrad U. Förstner and Till Sauerwein.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.8.5.
    </div>


</body>
</html>